src/System.Text.Utf8String/System/Text/Utf8String.cs

```diff
@@ -12,29 +12,45 @@
 namespace System.Text.Utf8
 {
     [DebuggerDisplay("{ToString()}u8")]
-    public partial ref struct Utf8String
+    public partial class Utf8String
     {
-        private readonly ReadOnlySpan<byte> _buffer;
+        private readonly byte[] _buffer;
 
         private const int StringNotFound = -1;
 
-        static Utf8String s_empty => default;
+        static Utf8String s_empty = new Utf8String(string.Empty);
 
         // TODO: Validate constructors, When should we copy? When should we just use the underlying array?
         // TODO: Should we be immutable/readonly?
         public Utf8String(ReadOnlySpan<byte> buffer)
         {
-            _buffer = buffer;
+            _buffer = buffer.ToArray();
         }
 
         public Utf8String(byte[] utf8bytes)
         {
-            _buffer = new ReadOnlySpan<byte>(utf8bytes);
+            _buffer = utf8bytes;
         }
 
         public Utf8String(byte[] utf8bytes, int index, int length)
         {
-            _buffer = new ReadOnlySpan<byte>(utf8bytes, index, length);
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException("index");
+            }
+
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException("length");
+            }
+
+            if (index + length > utf8bytes.Length)
```

Comment:
index + length could result in integer overflow

Consider something like the following for the bounds checks:
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Span.cs#L58-L63
```C#
            if (utf8bytes == null)
                // throw null exception
            if ((uint)index > (uint)utf8bytes.Length || (uint)length > (uint)(utf8bytes.Length - index))
                // throw out of range exception
```